### ISSUE TITLE
Problems using opton "historyLastStartDate" and "historyLastEndDate"

### ISSUE THREAD

**Issue by dg0815** [PLCHome/growatt#5]
Hello,
I am an absolute beginner with npm and JavaScript. My intention is to read the history data of my Growatt inverter from Growatt Webserver.  I installed npm on my Windows computer and implemented your example  test script. It worked fine! Afterwords I tried tu use the options "historyLastStartDate" and "historyLastEndDate" in order to read history data of a defined time period. But when I define options like this:
let options = {plantData:false,deviceData:true,deviceTyp:true,weather:false,historyLastStartDate:"2022-03-26"}
I get the following error messages:

![grafik](https://user-images.githubusercontent.com/102605306/160640585-dd94d8eb-9d13-4020-8d63-000d09f62e4e.png)

**Comment by PLCHome** [comment]
it mus be an date object not an String:

`,historyLastStartDate:new Date(2022, 3, 26)`

or for now:
`,historyLastStartDate:new Date()`

https://www.w3schools.com/jsref/jsref_obj_date.asp   ==> constructor

For Beginners ist maybe simpler to use ioBroker..

**Comment by dg0815** [comment]
Oh, thank you very much for your really quick response. Now I tried "
,historyLastStartDate:new Date(2022, 03, 26)
 and the following error message is the result:
![grafik](https://user-images.githubusercontent.com/102605306/160646125-984321d5-4931-4aaa-930a-12afcc692266.png)

I tried ioBroker a few days ago, but it seems to read always only ONE data set. I hoped that your options  "historyLastStartDate" and "historyLastEndDate" allow to read the inverter data for a whole week for example. Will this work?

**Comment by PLCHome** [comment]
kidding?
```
> let options = {plantData:false,deviceData:true,deviceTyp:true,weather:false,historyLastStartDate:new Date(2022, 3, 26)}
undefined
> options
{
  plantData: false,
  deviceData: true,
  deviceTyp: true,
  weather: false,
  historyLastStartDate: 2022-04-25T22:00:00.000Z
}
>
```

**Comment by dg0815** [comment]
Really not kidding. See screenshot. Could there be a problem with Windows Visual Studio Code?

![grafik](https://user-images.githubusercontent.com/102605306/160875032-a07f01c3-f145-498f-8120-910189aa001c.png)

**Comment by PLCHome** [comment]
Leading zeros must not be used with `'use strict';` : 03 ==> 3

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode

> Fifth, a strict mode in ECMAScript 5 [forbids a 0-prefixed octal literal or octal escape sequence](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Deprecated_octal). Outside strict mode, a number beginning with a 0, such as 0644, is interpreted as an octal number (0644 === 420), if all digits are smaller than 8. Octal escape sequences, such as "\45", which is equal to "%", can be used to represent characters by extended-ASCII character code numbers in octal. In strict mode, this is a syntax error. In ECMAScript 2015, octal literals are supported by prefixing a number with "0o"; for example:
> 
> `var a = 0o10; // ES2015: Octal`
> 

more from Date();

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date

**Comment by dg0815** [comment]
Thank you very much, you are right, "historyLastStartDate:new Date(2022, 2, 26)" will produce no error. 

But the options historyLastStartDate and historyLastEndDate do not work as I expected. The output always is only ONE dataset and not the complete dataset of the given time period. For exmaple
"historyLastStartDate:new   Date(2022,2,19),historyLastEndDate:new Date(2022,2,22)" 
results in the dataset from "2022-03-21T17:42:56.000Z" which is obviously the very last dataset in this time period. 
I cannot retrieve any other dataset, even if I define hours and minutes as well, e. g. "historyLastStartDate:new Date(2022,2,21,1,0,0),historyLastEndDate:new Date(2022,2,21,12,0,0)", I get back the dataset from "2022-03-21T17:42:56.000Z". I at least expected to get the "historyLast" of the last dataset in this time period.

**Comment by PLCHome** [comment]
history Last means the last dataset from the history, or the youngest.

The method was developed to retrieve the current performance data from a MAX. First we tried this over the chart but it didn't work. 
```resolve(res.data.obj.datas[0]);```

`datas[0]` means the first data.

I can have a look what I can do.

**Comment by PLCHome** [comment]
New options: 

historyAll: 	All data records from the system history table is called up here.
historyStart:       The server does not send all data for the time range. With the starting index you get the next rows.

**Comment by dg0815** [comment]
Hello, your new options sound great!!!
But to be honest: I have no idea how to use them. I tried several combinations with historyAll, historyLastStartDate, historyLastEndDate, historyStart  but I never got back any history data. And: "historyLast" is no more displayed, even if I just leave options={}. 
How can I retrieve history data now?

**Comment by PLCHome** [comment]
Adding object attribute:

```
let options={plantData:false,deviceData:true,deviceTyp:true,weather:false,historyLastStartDate:new Date(2022, 3, 26)};
options.historyAll = true
```
or  in json
```
let options = {plantData:false,deviceData:true,deviceTyp:true,weather:false,historyLastStartDate:new Date(2022, 3, 26), historyAll : true};
```

did you upgrade the growatt node in your project?

**Comment by dg0815** [comment]
Hello, yes I did upgrade growatt in my project with the commands
npm uninstall growatt
npm install growatt
The command "npm list" showed "growatt@0.2.8"

After a lot of debugging I found that in "getAllPlantDeviceData" you did a big change of your code after the line
`const device = await this.getDevicesByPlant(plantId).catch(e => {`
In my environment the return value of "device" is "device.obj.inv" with the following values:
device.obj.inv
(1) [Array(3)]
0: (3) ['EHE3A21030', 'Dachanlage', '0']
length: 1

Maybe you don't handle "device.obj.inv" in your new code using the variable "growattTyp"?

**Comment by PLCHome** [comment]
inv was broken i fixed it last night

**Comment by dg0815** [comment]
Thanks a lot, now it worked!
I had to modify your statement
`res.data.obj.datas.forall(data => this.correctTime(data));`
in getHistory because I've got the error message forall: undefined function. I modified it to
`res.data.obj.datas.forEach(data => this.correctTime(data));`
and now it works fine. 
I found that the array that is returned by getHistory is limit to 80 elements. So I will have to play with various StartDates and EndDates to get the history data of a whole time period. 
Thank your for your great support!!!
